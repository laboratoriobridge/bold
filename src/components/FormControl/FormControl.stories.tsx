import { boolean, text } from '@storybook/addon-knobs'
import React from 'react'

import { TextInput } from '../TextField'

import { FormControl } from './FormControl'

export default {
  title: 'Components/FormControl',
}

export const Default = () => (
  <FormControl
    label={text('label', 'Label')}
    htmlFor={text('id', 'input-id')}
    error={text('error', '')}
    inline={boolean('inline', false)}
    required={boolean('required', true)}
  >
    <TextInput id={text('id', 'input-id')} />
  </FormControl>
)
