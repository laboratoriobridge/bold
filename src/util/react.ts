import { MutableRefObject, Ref } from 'react'

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}
