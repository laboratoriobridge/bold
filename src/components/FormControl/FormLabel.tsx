import React, { CSSProperties } from 'react'

import { useLocale } from '../../i18n'
import { ExternalStyles, Theme, useStyles } from '../../styles'
import { Omit } from '../../util/types'

export interface FormLabelProps extends Omit<React.LabelHTMLAttributes<HTMLLabelElement>, 'style'> {
  label: React.ReactNode
  required?: boolean
  style?: ExternalStyles
  markerStyle?: ExternalStyles
}

export function FormLabel(props: FormLabelProps) {
  const { label, required, style, markerStyle, ...rest } = props
  const { classes, css } = useStyles(createStyles)
  const locale = useLocale()

  return (
    <label className={css(classes.label, style)} {...rest}>
      {label}

      {required && (
        <span title={locale.formControl.required} className={css(classes.marker, markerStyle)}>
          &#42;
        </span>
      )}
    </label>
  )
}

export const createStyles = (theme: Theme) => ({
  label: {
    fontWeight: 'bold',
  } as CSSProperties,
  marker: {
    color: theme.pallete.status.danger.main,
    marginLeft: '0.25rem',
  } as CSSProperties,
})
