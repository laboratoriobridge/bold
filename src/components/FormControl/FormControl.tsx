import React, { CSSProperties } from 'react'

import { VFlow } from '../VFlow'
import { ExternalStyles, useStyles } from '../../styles'

import { FormError } from './FormError'
import { FormLabel, FormLabelProps } from './FormLabel'

export interface FormControlProps {
  htmlFor?: string
  label?: FormLabelProps['label']
  labelId?: FormLabelProps['id']
  errorId?: string
  required?: FormLabelProps['required']
  error?: React.ReactNode
  inline?: boolean
  children?: React.ReactNode
  style?: ExternalStyles
}

export function FormControl(props: FormControlProps) {
  const { children, htmlFor, error, label, labelId, errorId, inline, required, style } = props
  const { classes, css } = useStyles(createStyles)

  const isInline = inline && label

  return (
    <VFlow vSpacing={0}>
      <div className={css([classes.formControl, isInline && classes.formControlInline, style])}>
        {label && (
          <FormLabel
            id={labelId}
            required={required}
            style={css(classes.label, isInline && classes.labelInline)}
            markerStyle={isInline && classes.markerInline}
            htmlFor={htmlFor}
            label={label}
          />
        )}
        {children}
      </div>
      {error && (
        <FormError id={errorId} style={classes.error} role='alert'>
          {error}
        </FormError>
      )}
    </VFlow>
  )
}

const createStyles = () => ({
  formControl: {} as CSSProperties,
  formControlInline: {
    display: 'grid',
    gridAutoFlow: 'column',
    gridTemplateColumns: 'min-content',
    alignItems: 'baseline',
    gridGap: '1rem',
  } as CSSProperties,
  label: {
    display: 'block',
    marginBottom: '0.25rem',
    lineHeight: '20px',
  } as CSSProperties,
  labelInline: {
    position: 'relative',
    whiteSpace: 'nowrap',
    marginBottom: 0,
  } as CSSProperties,
  markerInline: {
    position: 'absolute',
  } as CSSProperties,
  error: {
    marginTop: '0.25rem',
    lineHeight: '20px',
  } as CSSProperties,
})
