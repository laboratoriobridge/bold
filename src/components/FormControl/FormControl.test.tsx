import { render } from '@testing-library/react'
import React from 'react'

import { FormControl } from './FormControl'

it('should render correctly', () => {
  const { container, rerender } = render(<FormControl>Test</FormControl>)
  expect(container).toMatchSnapshot()

  rerender(
    <FormControl label='Label' error='Error' htmlFor='test' required>
      Test
    </FormControl>
  )
  expect(container).toMatchSnapshot()
})

it('should accept the style prop', () => {
  const { container } = render(<FormControl style={{ color: 'pink' }}>Test</FormControl>)
  expect(container).toMatchSnapshot()
})
