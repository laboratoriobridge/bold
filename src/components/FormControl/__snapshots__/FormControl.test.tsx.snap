// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should accept the style prop 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  display: inline-block;
  margin-bottom: 0rem;
  margin-top: 0rem;
}

.emotion-1:first-of-type {
  margin-top: 0;
}

.emotion-1:last-of-type {
  margin-bottom: 0;
}

.emotion-1:empty {
  display: none;
}

.emotion-0 {
  color: pink;
}

<div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-0"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  display: inline-block;
  margin-bottom: 0rem;
  margin-top: 0rem;
}

.emotion-1:first-of-type {
  margin-top: 0;
}

.emotion-1:last-of-type {
  margin-bottom: 0;
}

.emotion-1:empty {
  display: none;
}

<div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-0"
      >
        Test
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly 2`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-3 {
  display: inline-block;
  margin-bottom: 0rem;
  margin-top: 0rem;
}

.emotion-3:first-of-type {
  margin-top: 0;
}

.emotion-3:last-of-type {
  margin-bottom: 0;
}

.emotion-3:empty {
  display: none;
}

.emotion-0 {
  color: #D01E29;
  margin-left: 0.25rem;
}

.emotion-1 {
  font-weight: bold;
  display: block;
  margin-bottom: 0.25rem;
  line-height: 20px;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #D01E29;
  margin-top: 0.25rem;
  line-height: 20px;
}

<div>
  <div
    class="emotion-6"
  >
    <div
      class="emotion-3"
    >
      <div
        class="emotion-2"
      >
        <label
          class="emotion-1"
          for="test"
        >
          Label
          <span
            class="emotion-0"
            title="Required field"
          >
            *
          </span>
        </label>
        Test
      </div>
    </div>
    <div
      class="emotion-3"
    >
      <div
        class="emotion-4"
        role="alert"
      >
        Error
      </div>
    </div>
  </div>
</div>
`;
